File: ext/encryption.js
/*
@fileoverview Node.JS Blowfish encryption.
@author David Parlevliet
@version 20130305
@preserve Copyright 2013 David Parlevliet.

Encryption
==========
Blowfish encryption. Tested Decypherable between PHP<->Node.JS<->Python
*/
var crypto = require('crypto');
var algorithm = "bf-ecb";

function pad(text) {
  var pad_bytes = 8 - (text.length % 8)
  for (var x=1; x<=pad_bytes;x++)
    text = text + String.fromCharCode(0)
  return text;
}

function Encryption() {
  self = this;
  self.encrypt = function(data, key) {
    var cipher = crypto.createCipheriv(algorithm, Buffer(key), '');
    cipher.setAutoPadding(false);
    try {
      return Buffer(cipher.update(pad(data),
                      'utf8', 'binary') + cipher.final('binary'), 'binary')
                        .toString('hex');
    } catch (e) {
      console.log('Encryption error: ', e, data, key);
      return null;
    }
  }

  self.decrypt = function(data, key) {
    var decipher = crypto.createDecipheriv(algorithm, Buffer(key), '');
    decipher.setAutoPadding(false);
    try {
      return (decipher.update(Buffer(data, 'hex').toString('binary'),
              'binary', 'utf8') + decipher.final('utf8')).replace(/\x00+$/g, '');
    } catch (e) {
      console.log('Decryption error: ', e, data, key);
      return null;
    }
  }
}

module.exports = Encryption;

File: ext/encryption.py
"""
@fileoverview Python Blowfish encryption.
@author David Parlevliet
@version 20130305
@preserve Copyright 2013 David Parlevliet.

Encryption
==========
Blowfish encryption. Tested Decypherable between PHP<->Node.JS<->Python
"""
from Crypto.Cipher import Blowfish
from base64 import b64encode, b64decode
from django.conf import settings
from random import randrange
from binascii import unhexlify, hexlify


def Encrypt(text, key):
  pad_bytes = 8 - (len(text) % 8)
  for i in range(pad_bytes):
    text += chr(0)
  cipher = Blowfish.new(key, Blowfish.MODE_ECB, '')
  return str(cipher.encrypt(text)).encode("hex")


def Decrypt(text, key):
  try:
    cipher = Blowfish.new(key, Blowfish.MODE_ECB, '')
    output = cipher.decrypt(text.decode("hex"))
    # Trim off extra nulls
    while output[-1] == chr(0):
      output = output[:-1]
    return output
  except:
    return None
